cuppa/src/main/java/com/hartwig/hmftools/cup/cli/ShellCommand.java

```diff
@@ -68,77 +68,56 @@ public ShellCommand timeout(long timeout)
         return this;
     }
 
-    public String toString()
-    {
-        return mProcessBuilder.command().toString();
-    };
-
     public ShellCommand run()
     {
         checkComplete();
-
         try
         {
             if(mShowCommand)
-                CUP_LOGGER.info("Running command: " + this);
+                CUP_LOGGER.info("Running command: [{}]", this);
 
             Process process = mProcessBuilder
                     .redirectErrorStream(true) // Merge stdout and stderr
                     .start();
 
-            mStdout = captureStdout(process, mLogLevel);
-
             mComplete = process.waitFor(mTimeout, mTimeoutUnit);
-            process.destroy();
-
+            mStdout = captureStdout(process, mLogLevel);
             mExitCode = process.exitValue();
-            if(mExitCode > 0)
-                throw new RuntimeException();
         }
         catch(InterruptedException | IllegalThreadStateException e)
         {
             CUP_LOGGER.error("Failed to run command({}) due to time out after {} {}", mProcessBuilder.command(), mTimeout, mTimeoutUnit.toString().toLowerCase());
-            System.exit(mExitCode);
+            mExitCode = 1;
         }
         catch(IOException | RuntimeException e)
         {
-            CUP_LOGGER.error("Failed to run command({}) with exit code({})", mProcessBuilder.command(), String.valueOf(mExitCode));
-            System.exit(mExitCode);
+            CUP_LOGGER.error("Failed to run command({})", mProcessBuilder.command());
+            mExitCode = 1;
         }
 
         return this;
     }
 
-    private static List<String> captureStdout(Process process, Level logLevel)
-    {
+    private static List<String> captureStdout(Process process, Level logLevel) {
         List<String> stdout = new ArrayList<>();
-
-        // Use a separate thread to avoid reader blocking the main thread (which would cause the process to never finish)
-        Thread thread = new Thread(() -> {
```

Comment:
`captureStdOut` was in it's own thread to avoid delayed printing of python log messages - they were only printed once the shell process was finished (or crashed). 

If you don't think this is important, we can use with your simplified implementation.

Reply:
I think more important is whether you think it's an issue. I always prefer a simplified implementation and admit I modified this while working with test data which probably completes much faster than a "real" patient.

I'm not convinced that the threading doesn't have something to do with the intermittent hangs you mentioned; one way to proceed would be to use this for a while and see if they go away and/or you don't notice the "banking up" of the messages.

Additionally I think you can potentially use the `inheritIO` method of the process builder to make the spawned process use the same streams as the parent (Java) process, which I think would allow the output to appear immediately without an extra thread. This however might not be exactly what you want. Maybe worth discussing further.

Reply:
Thanks a lot for the `inheritIO` tip. I've now implemented this and the log messages are no longer delayed, and i could remove the redundant java log level too :)

```
13:25:29.063 [main] [INFO ] Cuppa version 2.2.1
13:25:29.065 [main] [INFO ] Creating output dir: /var/folders/r7/3t6mpsqs2ysfqk1pslgsmy8h0000gn/T/cuppa_output
13:25:29.272 [main] [INFO ] Writing data to: /var/folders/r7/3t6mpsqs2ysfqk1pslgsmy8h0000gn/T//cuppa_output//TUMOR_SAMPLE.cuppa_data.tsv.gz
13:25:29.285 [main] [INFO ] Cuppa data extraction complete, mins(0.004)
13:25:29.286 [main] [INFO ] Predicting using command: /bin/bash -c /Users/lnguyen/.pyenv/versions/3.9.4/envs/pycuppa_venv/bin/python -m cuppa.predict --classifier_path /Users/lnguyen/Hartwig/common-resources-public/cuppa/37/cuppa_classifier.37.pickle.gz --features_path /var/folders/r7/3t6mpsqs2ysfqk1pslgsmy8h0000gn/T//cuppa_output//TUMOR_SAMPLE.cuppa_data.tsv.gz --output_dir /var/folders/r7/3t6mpsqs2ysfqk1pslgsmy8h0000gn/T//cuppa_output/ --sample_id TUMOR_SAMPLE --log_format '%(asctime)s.%(msecs)03d [python] %(levelname)s %(name)s.%(funcName)s | %(message)s'
13:25:29.286 [main] [INFO ] Running command: [/bin/bash -c /Users/lnguyen/.pyenv/versions/3.9.4/envs/pycuppa_venv/bin/python -m cuppa.predict --classifier_path /Users/lnguyen/Hartwig/common-resources-public/cuppa/37/cuppa_classifier.37.pickle.gz --features_path /var/folders/r7/3t6mpsqs2ysfqk1pslgsmy8h0000gn/T//cuppa_output//TUMOR_SAMPLE.cuppa_data.tsv.gz --output_dir /var/folders/r7/3t6mpsqs2ysfqk1pslgsmy8h0000gn/T//cuppa_output/ --sample_id TUMOR_SAMPLE --log_format '%(asctime)s.%(msecs)03d [python] %(levelname)s %(name)s.%(funcName)s | %(message)s']
13:25:30.182 [python] [INFO ] CuppaFeaturesLoader._load_data | Loading single sample data file: /var/folders/r7/3t6mpsqs2ysfqk1pslgsmy8h0000gn/T//cuppa_output//TUMOR_SAMPLE.cuppa_data.tsv.gz
13:25:30.195 [python] [DEBUG] CuppaFeaturesLoader._print_stats | Loaded 6220 features from sample(TUMOR_SAMPLE)
13:25:30.196 [python] [INFO ] CuppaClassifier.from_file | Loading classifier from: /Users/lnguyen/Hartwig/common-resources-public/cuppa/37/cuppa_classifier.37.pickle.gz
...
```